BACKEND/app/crud/sources.py

```diff
@@ -0,0 +1,84 @@
+from sqlalchemy.orm import Session
+from app.models.sources import Source
+from app.models.languages import Language
+from app.models.versions import Version
+from app.schemas.sources import SourceCreate, SourceUpdate
+from fastapi import HTTPException, status
+from uuid import UUID
+
+class SourceService:
+    def create_source(self, db: Session, source_data: SourceCreate) -> Source:
+        # Fetch language
+        language = db.query(Language).filter(Language.language_id == source_data.language_id).first()
+        if not language:
+            raise HTTPException(
+                status_code=status.HTTP_404_NOT_FOUND,
+                detail=f"Language with ID {source_data.language_id} not found"
+            )
+
+        # Fetch version
+        version = db.query(Version).filter(Version.version_id == source_data.version_id).first()
+        if not version:
+            raise HTTPException(
+                status_code=status.HTTP_404_NOT_FOUND,
+                detail=f"Version with ID {source_data.version_id} not found"
+            )
+
+        # Create Source with auto-filled names
+        source = Source(
+            language_id=source_data.language_id,
+            language_name=language.name,
+            version_id=source_data.version_id,
+            version_name=version.version_name,
+            description=source_data.description,
+        )
+
+        db.add(source)
+        db.commit()
+        db.refresh(source)
+        return source
+
+    def get_source(self, db: Session, source_id: UUID) -> Source:
```

Comment:
you can rename this function name if its fetching by source_id. rename to get_source_id